shouldHaveLength should show the length on failure

Currently it will say

java.lang.AssertionError: qwerty should have length 32

Yep.

On Tue, 2 Jul 2019, 18:37 Gustavo Fão Valvassori, <notifications@github.com>
wrote:

> Like this: java.lang.AssertionError: hello should have length 32, but
> instead was 5?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kotlintest/kotlintest/issues/869?email_source=notifications&email_token=AAFVSGU2QWF3ZG2VZYHVS4DP5PRCXA5CNFSM4H46DJ4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZC2Y3Y#issuecomment-507882607>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFVSGVRELKJ75TNL4JVGTLP5PRCXANCNFSM4H46DJ4A>
> .
>

--- a/kotlintest-assertions/src/main/kotlin/io/kotlintest/matchers/StringMatchers.kt
+++ b/kotlintest-assertions/src/main/kotlin/io/kotlintest/matchers/StringMatchers.kt
@@ -45,6 +45,6 @@ fun strlen(length: Int): Matcher<String?> = haveLength(length)
 fun haveLength(length: Int) = neverNullMatcher<String> { value ->
   Result(
       value.length == length,
-      "${convertValueToString(value)} should have length $length",
+      "${convertValueToString(value)} should have length $length, but instead was ${value.length}",
       "${convertValueToString(value)} should not have length $length")
 }

--- a/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/string/StringMatchersTest.kt
+++ b/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/string/StringMatchersTest.kt
@@ -518,13 +518,13 @@ class StringMatchersTest : FreeSpec() {
         "hello" shouldHaveLength 5
         shouldThrow<AssertionError> {
           "" should haveLength(3)
-        }.message shouldBe "<empty string> should have length 3"
+        }.message shouldBe "<empty string> should have length 3, but instead was 0"
         shouldThrow<AssertionError> {
           "" shouldHaveLength 3
-        }.message shouldBe "<empty string> should have length 3"
+        }.message shouldBe "<empty string> should have length 3, but instead was 0"
         shouldThrow<AssertionError> {
           "hello" shouldHaveLength 3
-        }.message shouldBe "hello should have length 3"
+        }.message shouldBe "hello should have length 3, but instead was 5"
         shouldThrow<AssertionError> {
           "hello" shouldNotHaveLength 5
         }.message shouldBe "hello should not have length 5"

--- a/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/properties/shrinking/ShrinkTest.kt
+++ b/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/properties/shrinking/ShrinkTest.kt
@@ -87,7 +87,7 @@ class ShrinkTest : StringSpec({
       assertAll(gen) { a ->
         a.shouldHaveLength(10)
       }
-    }.message shouldBe "Property failed for\nArg 0: <empty string> (shrunk from asjfiojoqiwehuoahsuidhqweqwe)\nafter 1 attempts\nCaused by: asjfiojoqiwehuoahsuidhqweqwe should have length 10"
+    }.message shouldBe "Property failed for\nArg 0: <empty string> (shrunk from asjfiojoqiwehuoahsuidhqweqwe)\nafter 1 attempts\nCaused by: asjfiojoqiwehuoahsuidhqweqwe should have length 10, but instead was 28"
   }
 
   "should shrink strings to min failing size" {
@@ -100,6 +100,6 @@ class ShrinkTest : StringSpec({
       assertAll(gen) { a ->
         a.padEnd(10, '*').shouldHaveLength(10)
       }
-    }.message shouldBe "Property failed for\nArg 0: aaaaaaaaaaaaaa (shrunk from asjfiojoqiwehuoahsuidhqweqwe)\nafter 1 attempts\nCaused by: asjfiojoqiwehuoahsuidhqweqwe should have length 10"
+    }.message shouldBe "Property failed for\nArg 0: aaaaaaaaaaaaaa (shrunk from asjfiojoqiwehuoahsuidhqweqwe)\nafter 1 attempts\nCaused by: asjfiojoqiwehuoahsuidhqweqwe should have length 10, but instead was 28"
   }
 })

